src/weakpoint.cpp

```diff
@@ -0,0 +1,77 @@
+#include "weakpoint.h"
+
+#include <string>
+#include <utility>
+
+#include "assign.h"
+#include "creature.h"
+#include "damage.h"
+#include "item.h"
+#include "rng.h"
+
+class JsonArray;
+class JsonObject;
+
+weakpoint::weakpoint()
+{
+    // arrays must be filled manually to avoid UB.
+    armor_mult.fill( 1.0f );
+    armor_penalty.fill( 0.0f );
+}
+
+void weakpoint::load( const JsonObject &jo )
+{
+    assign( jo, "id", id );
+    assign( jo, "name", name );
+    assign( jo, "coverage", coverage, false, 0.0f, 100.0f );
+    if( jo.has_object( "armor_mult" ) ) {
+        armor_mult = load_damage_array( jo.get_object( "armor_mult" ), 1.0f );
+    }
+    if( jo.has_object( "armor_penalty" ) ) {
+        armor_penalty = load_damage_array( jo.get_object( "armor_penalty" ), 0.0f );
+    }
+    // Set the ID to the name, if not provided.
+    if( id.empty() ) {
+        id = name;
+    }
+}
+
+void weakpoint::apply_to( resistances &resistances ) const
+{
+    for( int i = 0; i < static_cast<int>( damage_type::NUM ); ++i ) {
+        resistances.resist_vals[i] *= armor_mult[i];
+        resistances.resist_vals[i] -= armor_penalty[i];
+    }
+}
+
+float weakpoint::hit_chance() const
+{
+    return coverage;
+}
+
+const weakpoint *weakpoints::select_weakpoint() const
+{
+    float idx = rng_float( 0.0f, 100.0f );
+    for( const weakpoint &weakpoint : weakpoint_list ) {
+        float hit_chance = weakpoint.hit_chance( );
+        if( hit_chance <= idx ) {
```

Comment:
So I'm trying to understand this section, if coverage (AKA hit chance) is 5 then I could roll 5 to 100 and 'select' the weak point. Isn't this backwards, shouldn't I have only a 5% chance to hit the weak point? 

Also wouldn't the order of weakpoints in JSON have a large effect on what gets selected? If something has a large change of being selected at the beginning, than nothing else will ever be selected.

Reply:
Looking at it some more, I think the condition should be `if( hit_chance > idx ) {` and then when loading the weak points, some math will have to be done to turn the percent chance of each weakpoint into a a range of changes.

As in, two 5 coverage weakpoints will lead to a the first with 5 and the second with 10, so a roll (0,5] will lead to the first weakpoint being chosen and a roll (5,10] will lead to the second chosen. While doing this math, it will probably be a good idea to do some kind of verification that the total combined coverages are <= 100

Reply:
I'm not sure what standard procedure is, so I made a PR for it. I'm going to try and make a test for it, but I haven't done that before. In any case, we should add some a percentage chance to the debug monster that isn't 50% :P

Reply:
Nice catch! The hit chance comparison should be flipped around.

If the probabilities sum to more than 100%, the order of the weak points does have an effect. Weak points earlier in the list will be preferred to weak points further down. This was intentional. In s future change, we will let character skill effect the hit chance. If a character is skilled enough to have a 100% weak point hit chance, they should be able to prioritize the more vulnerable weak points. In practice, I don't expect that to come up very often.

Because hit chances will be determined dynamically, there's not much benefit to constructing an explicit set of ranges. The current linear search method will be just as performant.